File: Problems/50.powx-n.cpp
// title: Pow(x, n)
// timestamp: 2024-02-22 07:51:36
// problemUrl: https://leetcode.com/problems/powx-n/
// memory: 8.2 MB
// runtime: 0 ms

// #include <cmath>

class Solution {
public:
    double myPow(double x, int n) {
        if (x == 0) {
            return 0;
        }
        if (n == 0) {
            return 1;
        }
        if (n < 0) {
            long long m = n;
            return pow(1/x, -m);
        }
        return pow(x, n);
    }
    double pow(double x, long long n) {
        if (n == 1) {
            return x;
        }
        if (n % 2 == 0) {
            return pow(x * x, n / 2);
        } else {
            return x * pow(x * x, n / 2);
        }
    }
};

File: Problems/50.powx-n.py
# title: Pow(x, n)
# timestamp: 2025-06-17 23:57:26
# problemUrl: https://leetcode.com/problems/powx-n/
# memory: 18 MB
# runtime: 0 ms

class Solution:
    def myPow(self, x: float, n: int) -> float:
        sign = 1
        res = 1
        if n < 0:
            sign = -1
            n = -n
        if n == 0:
            return 1
        
        while n:
            if n & 1:
                res *= x
            x *= x
            n >>= 1
        
        if sign == -1:
            return 1 / res
        return res